scriptworker/artifacts.py

```diff
@@ -306,3 +306,39 @@ def get_expiration_arrow(context):
 
     await raise_future_exceptions(tasks)
     return files
+
+
+def get_upstream_artifacts(context):
+    """List the downloaded upstream artifacts.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+
+    Returns:
+        dict: the paths of uploaded artifacts, sorted per task_id
+
+    Raises:
+        scriptworker.exceptions.ScriptWorkerTaskException: when an artifact doesn't exist.
+    """
+    task_ids_and_relative_paths = [
+        (artifact_definition['taskId'], artifact_definition['paths'])
+        for artifact_definition in context.task['payload']['upstreamArtifacts']
+    ]
+
+    return {
+        task_id: [
+            _get_single_upstream_artifact(context, task_id, path)
+            for path in paths
+        ]
+        for task_id, paths in task_ids_and_relative_paths
+    }
+
+
+def _get_single_upstream_artifact(context, task_id, path):
```

Comment:
I ended up not usgin with functools. Instead I used a regular method, and cot_dir is a property.